### ISSUE TITLE
Possible typing issue w/ Entity Service API & populated relations

### ISSUE THREAD

**Issue by sopwerdna** [strapi/strapi#19252]
### Required System information

- Node.js version: 20.8.0
- NPM version: 10.2.0
- Strapi version: 4.17.0
- Database: Postgresql
- Operating system: macOS, linux
- Is your project Javascript or Typescript: Typescript

### Describe the bug
We've run into a typescript error upon setting up the project with new versions after the 4.14.x typescript migration changes. In short, I believe this is a result of incorrect or conflicting types coming from calls to the Entity Service API, or objects returned from those calls. Unsure if this is a regression or simply an existing issue that was uncovered by the changes in 4.14.x.

### Steps to reproduce the behavior
1. Add a content type (posts for example, `api::post.post`)
2. Add a `oneToMany` relation from the built in Users table to that content type
3. In a custom controller, call the Strapi Entity Service [update()](https://docs.strapi.io/dev-docs/api/entity-service/crud#update) which states that updating relations follows the same syntax as the REST API and includes an [example of the syntax](https://docs.strapi.io/dev-docs/api/rest/relations#connect).
4. Using that syntax gives a type error when using typescript: ```Type '{ connect: ID[]; }' is not assignable to type 'XOneInput'. Object literal may only specify known properties, and 'connect' does not exist in type 'LongHand'.```

### Expected behavior
In previous versions, our input to the entity service `update(...., data: {...}...)` function was typed as `XManyInput` which allows partial updates and that `connect:` syntax. Now it's expecting `XOneInput` even though that relation is definitely a one to many relation and likely should be treated as an array. We're also seeing some other related issues (similar to #18784) which may be a result of the same type conflicts.

Looking closer at the types, I believe this is either a problem in the typescript types for Strapi's entity service API types (in the `@strapi/types` package), or possibly in the documentation (if there's really a different syntax now). It looks to me like the type on that particular one to many relation _should_ be `XManyInput` which allows partial updates and that `connect:` syntax.

### Additional context

For further context, [this](https://github.com/strapi/strapi/blob/develop/packages/core/types/src/modules/entity-service/params/attributes/relation.ts#L23) seems to be the file in the strapi types package with the problematic types
```type Set = { set: ShortHand[] | LongHand[] | null };
type Connect = { connect: ShortHand[] | WithPositionArguments<LongHand>[] };
type Disconnect = { disconnect: ShortHand[] | LongHand[] };
type FullUpdate = Set;
type PartialUpdate = Partial<Connect & Disconnect>;
type XOneInput = ShortHand | LongHand | null;
type XManyInput = ShortHand[] | LongHand[] | null | PartialUpdate | FullUpdate;
type RelationInputValue<TRelationKind extends Attribute.RelationKind.Any> = Utils.Expression.If<
  Attribute.IsManyRelation<TRelationKind>,
  XManyInput,
  XOneInput
>;
```

Thanks in advance for your help in tracking down the problem here, and fixing the issue if it does turn out to be a problem with Strapi's types. As always, let me know if I can provide any further information or details on the problem here.

**Comment by hussain-mustafa990** [comment]
I have also experience this issue on version 4.17.1

**Comment by williamatpaper** [comment]
Experiencing this issue on 4.20.4

**Comment by MarijnFK** [comment]
Same / related issue:
![image](https://github.com/strapi/strapi/assets/47977238/3e8f7764-96ca-4760-89b0-9ffdec367f12)

Workaround is to use the query engine, but that does not work when creating / adding components in a repeatable component field

**Comment by MarijnFK** [comment]
> @MarijnFK would you mind sharing the definition for the task-status content-type? I would like to try and reproduce the issue.

sure, but it's quite big. The problem also exists when just creating a content-type with just one field (i.e. 'visible')
```
{
  "kind": "collectionType",
  "collectionName": "task_statuses",
  "info": {
    "singularName": "task-status",
    "pluralName": "task-statuses",
    "displayName": "TaskStatus",
    "description": ""
  },
  "options": {
    "draftAndPublish": false
  },
  "pluginOptions": {},
  "attributes": {
    "task": {
      "type": "relation",
      "relation": "oneToOne",
      "target": "api::event-task.event-task"
    },
    "event": {
      "type": "relation",
      "relation": "oneToOne",
      "target": "api::event.event"
    },
    "team": {
      "type": "relation",
      "relation": "oneToOne",
      "target": "api::team.team"
    },
    "visible": {
      "type": "boolean",
      "default": true
    },
    "map_visible": {
      "type": "boolean",
      "default": true
    },
    "available": {
      "type": "boolean",
      "default": true
    },
    "unlocked": {
      "type": "boolean",
      "default": true
    },
    "status": {
      "type": "enumeration",
      "enum": [
        "none",
        "pending",
        "warning",
        "correct",
        "failed"
      ],
      "default": "none"
    },
    "answers": {
      "type": "relation",
      "relation": "oneToMany",
      "target": "api::answer.answer",
      "mappedBy": "taskStatus"
    },
    "hints": {
      "type": "component",
      "repeatable": true,
      "component": "games.consumed-hint"
    }
  }
}
```

**Comment by Convly** [comment]
I failed to reproduce the issue using the latest version (4.24.3).
![image](https://github.com/strapi/strapi/assets/25851739/65add139-22ea-4d92-a4c5-18f6397d4f8b)
Do you have any additional custom config that could impact this?

**Comment by MarijnFK** [comment]
i'm trying to reproduce the error in a clean install

**Comment by cyp3rius** [comment]
It's a dependencies conflict, please use `2.5.0` which is aligned to one of the recent versions of Strapi.

https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/issues/431#issuecomment-2128894524

**Comment by Convly** [comment]
> @Convly @derrickmehaffy It does not appear you followed the reproduction steps in the original bug report. I still encounter the same problem with `connect:` syntax as reported in the original bug report on version 4.24.3. The reproduction attempt in your picture does not seem to include a relation in the data model, which was the trigger for the original problem myself and @williamatpaper encountered.

I've now tried adding `oneToOne`, `oneToMany`, `manyToOne`, and `manyToMany` relations and haven't hit the issue.

However, after playing with removing node_modules, types, then re-installing/generating everything, I ended up with the "Object literal may only specify known properties..." error (even without the mentioned plugin installed)

I've reloaded my vs-code and the error went away, and I now have access to everything again. Could it be that the error comes from an invalid vscode state?

Do you have compilation errors when building/starting your application? Have you tried restarting/reloading your vscode?

**Comment by Convly** [comment]
Thanks for the reproduction repo, it helps a lot!

I can indeed see the issue (although I don't think https://github.com/strapi/strapi/issues/19252#issuecomment-2124234183 is related)

![image](https://github.com/strapi/strapi/assets/25851739/0744cbd7-8eda-46e4-a3ee-cc76a81b5b68)

The only problem I can see here is the longhand notation with positional arguments that needs to have `position` as optional instead of required.

Once the `position` property is added (or removed from the interface), it correctly infers the populated fields for the query and everything seems to work well.

![image](https://github.com/strapi/strapi/assets/25851739/36966d6b-a2e5-4aec-9bd7-661927c922be)

**Comment by MarijnFK** [comment]
@Convly it might not be related indeed. Sofar i have tested several repo's with the navigation-plugin and removed it (and clean cache, re-gen types), then it works again. Re-installing the plugin (again, clear cache and re-gen types) problem re-occurs. 

@cyp3rius can you verify if the problem could be related to the user-permissions plugin?

**Comment by Convly** [comment]
> I am curious, is this a documentation issue (i.e. the position argument is meant to be required in the long hand version) or is the fix to update the types to reflect the documentation? Either way I think this resolves my immediate problem. Thank you!

This is a bug, not a documentation issue, it's been fixed in #20370 and will be shipped Wednesday in v4.24.4

Sorry for the time it took to track down the issue, and thanks for the reproduction repo :slightly_smiling_face:

**Comment by Convly** [comment]
> @Convly it might not be related indeed. Sofar i have tested several repo's with the navigation-plugin and removed it (and clean cache, re-gen types), then it works again. Re-installing the plugin (again, clear cache and re-gen types) problem re-occurs.
> 
> @cyp3rius can you verify if the problem could be related to the user-permissions plugin?

Hey, since the issue seems to be unrelated to the original one, could you open a new one if you wish to continue the discussion on your problem? Thanks!

(in your new issue it would be awesome if you could include a reproduction repo or example)
